application/codeaction/codeaction.go

```diff
@@ -0,0 +1,90 @@
+package codeaction
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+
+	"github.com/snyk/snyk-ls/application/server/lsp"
+	"github.com/snyk/snyk-ls/domain/ide/converter"
+	"github.com/snyk/snyk-ls/domain/snyk"
+	"github.com/snyk/snyk-ls/internal/uri"
+)
+
+// IssuesProvider is an interface that allows to retrieve issues for a given path and range.
+// This is used instead of any concrete dependency to allow for easier testing and more flexibility in implementation.
+type issuesProvider interface {
+	IssuesFor(path string, r snyk.Range) []snyk.Issue
+}
+
+// CodeActionsService is an application-layer service for handling code actions.
+type CodeActionsService struct {
+	IssuesProvider issuesProvider
+
+	// issuesCache holds all the issues that were returns by the GetCodeActions method.
+	// This is used to resolve the code actions later on in ResolveCodeAction.
+	issuesCache map[uuid.UUID]cachedIssue
+	logger      zerolog.Logger
+}
+
+type cachedIssue struct {
+	issue  snyk.Issue
+	action snyk.CodeAction
+}
+
+func NewService(provider issuesProvider) CodeActionsService {
+	return CodeActionsService{
+		IssuesProvider: provider,
+		issuesCache:    make(map[uuid.UUID]cachedIssue),
+		logger:         log.With().Str("service", "CodeActionsService").Logger(),
+	}
+}
+
+func (c *CodeActionsService) GetCodeActions(params lsp.CodeActionParams) []lsp.CodeAction {
+	c.logger.Info().Msg("Received code action request")
+	path := uri.PathFromUri(params.TextDocument.URI)
+	r := converter.FromRange(params.Range)
+	issues := c.IssuesProvider.IssuesFor(path, r)
+	actions := converter.ToCodeActions(issues)
+
+	// The cache is cleared every time GetCodeActions is called, because the assumed workflow is:
+	// 1. User gets multiple code action options for a given path/range via textDocument/codeAction
+	// 2. User selects an action and the action is resolved via codeAction/resolve
+	// So there is no reason to store issues for longer than that.
+	for key := range c.issuesCache {
+		delete(c.issuesCache, key)
+	}
+	for _, issue := range issues {
+		for _, action := range issue.CodeActions {
+			if action.Uuid != nil {
+				cached := cachedIssue{
+					issue:  issue,
+					action: action,
+				}
+				c.issuesCache[*action.Uuid] = cached
+			}
+		}
+	}
+
+	c.logger.Info().Msg(fmt.Sprint("Returning ", len(actions), " code actions"))
+	return actions
+}
+
+func (c *CodeActionsService) ResolveCodeAction(action lsp.CodeAction) lsp.CodeAction {
+	c.logger.Info().Msg("Received code action resolve request")
+	t := time.Now()
+	cached := c.issuesCache[uuid.UUID(*action.Data)]
```

Comment:
Yeah, I don't think that we should mix code actions with issues - they are different entities, and code action is not a sub-entity of issue, and vice versa. They exist next to each other, but not with a must-have relation.